File: 2024/Day2/2.2.py
def isLineValid(numbers):
        increasing = False
        decreasing = False
        valid_difference = True

        for i in range(1, len(numbers)):
            diff = numbers[i] - numbers[i - 1]
            if not (1 <= abs(diff) <= 3):
                valid_difference = False
                break
        
            if numbers[i] > numbers[i - 1]:
                    increasing = True
            elif numbers[i] < numbers[i - 1]:
                    decreasing = True
        
        if (increasing != decreasing) and valid_difference:
              return True
        else:
              return False

with open('input.txt', 'r') as file:
    lines = file.readlines()

    count = 0

    for line in lines:
        numbers = list(map(int, line.split()))
        
        if(isLineValid(numbers)):
            count += 1
            continue

        for i in range(len(numbers)):
            modified_numbers = numbers[:i] + numbers[i + 1:] 
            if isLineValid(modified_numbers):
                count += 1 
                break 
        

    print(count)

File: 2025/Day2/2.2.swift
import Foundation

let fileURL = URL(fileURLWithPath: "input.txt")

func isInvalidRepeatedID(_ id: String) -> Bool {
    let doubled = id + id
    let trimmed = doubled.dropFirst().dropLast()
    return trimmed.contains(id)
}

do {
    let contents = try String(contentsOf: fileURL, encoding: .utf8)
    let lines = contents.split(separator: ",") 


    var mirroredNumbers: Set<Int> = []

    for line in lines {
        let rangeParts = line.split(separator: "-")
        
        guard rangeParts.count == 2 else {
            print("Unable to parse line: \(line)")
            continue 
        }

        if let bottomRange = Int(rangeParts[0]), let topRange = Int(rangeParts[1]) {
            Array(bottomRange ... topRange).forEach { number in  
                if isInvalidRepeatedID(String(number)) {
                    mirroredNumbers.insert(number)
                }
            }
        }
    }

    var total = 0
        print("mirroed numbers: \(mirroredNumbers)")
    for number in mirroredNumbers {
        total += number
    }
    print("Total: \(total)")

} catch {
    print("Error reading file: \(error)")
}